mindone/diffusers/models/autoencoders/vae.py

```diff
@@ -306,6 +307,371 @@ def construct(
+                prev_output_channel=None,
+                add_upsample=not is_final_block,
+                resnet_eps=1e-6,
+                resnet_act_fn=act_fn,
+                resnet_groups=norm_num_groups,
+                attention_head_dim=output_channel,
+                temb_channels=temb_channels,
+                resnet_time_scale_shift=norm_type,
+            )
+            self.up_blocks.append(up_block)
+            prev_output_channel = output_channel
+        self.up_blocks = nn.CellList(self.up_blocks)
+
+        # condition encoder
+        self.condition_encoder = MaskConditionEncoder(
+            in_ch=out_channels,
+            out_ch=block_out_channels[0],
+            res_ch=block_out_channels[-1],
+        )
+
+        # out
+        if norm_type == "spatial":
+            self.conv_norm_out = SpatialNorm(block_out_channels[0], temb_channels)
+        else:
+            self.conv_norm_out = GroupNorm(num_channels=block_out_channels[0], num_groups=norm_num_groups, eps=1e-6)
+        self.conv_act = SiLU()
+        self.conv_out = nn.Conv2d(block_out_channels[0], out_channels, 3, pad_mode="pad", padding=1, has_bias=True)
+
+        self.gradient_checkpointing = False
+
+    def construct(
+        self,
+        z: ms.Tensor,
+        image: Optional[ms.Tensor] = None,
+        mask: Optional[ms.Tensor] = None,
+        latent_embeds: Optional[ms.Tensor] = None,
+    ) -> ms.Tensor:
+        r"""The forward method of the `MaskConditionDecoder` class."""
+        sample = z
+        sample = self.conv_in(sample)
+
+        # middle
+        sample = self.mid_block(sample, latent_embeds)
+
+        # condition encoder
+        im_x = None
+        if image is not None and mask is not None:
+            masked_image = (1 - mask) * image
+            im_x = self.condition_encoder(masked_image, mask)
+
+        # up
+        for up_block in self.up_blocks:
+            if image is not None and mask is not None:
+                sample_ = im_x[str(tuple(sample.shape))]
+                mask_ = ops.interpolate(mask, size=sample.shape[-2:], mode="nearest")
+                sample = sample * mask_ + sample_ * (1 - mask_)
+            sample = up_block(sample, latent_embeds)
+        if image is not None and mask is not None:
+            sample = sample * mask + im_x[str(tuple(sample.shape))] * (1 - mask)
+
+        # post-process
+        if latent_embeds is None:
+            sample = self.conv_norm_out(sample)
+        else:
+            sample = self.conv_norm_out(sample, latent_embeds)
+        sample = self.conv_act(sample)
+        sample = self.conv_out(sample)
+
+        return sample
+
+
+class VectorQuantizer(nn.Cell):
+    """
+    Improved version over VectorQuantizer, can be used as a drop-in replacement. Mostly avoids costly matrix
+    multiplications and allows for post-hoc remapping of indices.
+    """
+
+    # NOTE: due to a bug the beta term was applied to the wrong term. for
```

Comment:
what bug?

Reply:
this comment comes from origin diffusers, not mindspore version 233